### ISSUE TITLE
crates.io versions 0.12 and 0.13 of amethyst don't compile on Mac OS

### ISSUE THREAD

**Issue by trsoluti** [amethyst/amethyst#2098]
## Description

The versions of Amethyst published on [crates.io](https://crates.io/crates/amethyst) don't compile on a Mac due to problems with metal v0.15.0 (from gfx-backend-metal 0.2.4, from rendy-util 0.3.1).

metal 0.15.0 fails with a number of occurrences of the following error:

```
error[E0282]: type annotations needed
...
consider giving 'result' a type
in this macro invocation
```

## Reproduction Steps

1. Create a new empty cargo project
2. add the dependency `amethyst = { version = "0.12.*", features = ["metal"] }` [or `0.13.2`]
3. `cargo build`

## What You Expected to Happen

All the Amethyst components should compile

## What Actually Happened

metal fails to compile.

## Screenshots and Logs

If applicable, add screenshots to help explain your problem. Any relevant logs would be helpful as well.

## Your Environment

1. OS: Mac OS Mojave
2. OS version: 10.14.16
3. Amethyst versions 0.12 and 0.13 (crates.io)

## Additional Context

The master version of Amethyst is using metal 0.16.0 (from gfx-backend-metal 0.3.3,
from rendy-util 0.4.1), which compiles.

**Comment by geoeo** [comment]
can confirm

**Comment by holmesmr** [comment]
I think this is a duplicate of this (includes a workaround): https://github.com/amethyst/amethyst/issues/2046

**Comment by geoeo** [comment]
Yes this fixes the compilation issue but effectively is a downgrade. Is this issue still being tracked?

**Comment by 0x6273** [comment]
This commit fixes it: https://github.com/gfx-rs/metal-rs/commit/032800fe731082ff115a5346b5178d12cfd1bf02#diff-b4aea3e418ccdb71239b96952d9cddb6

If metal can release a 0.15.1 with this commit, I think it would fix it.

**Comment by 0x6273** [comment]
Pretty much the only change in objc 0.2.7 was that it broke it's own api (something that patch versions **really** shouldn't do), so we could release new patch versions for 0.12 and 0.13 that pins objc to 0.2.6 if we can't get a 0.15.1 release of metal.

I made an issue asking for a 0.15.1 release here: https://github.com/gfx-rs/metal-rs/issues/110

Also, it would be great if someone with a Mac could verify that just applying the commit I mentioned on metal 0.15.0 really is enough to fix the issue.

**Comment by trsoluti** [comment]
I added the following lines to my Cargo.toml:

```rust
[patch.crates-io]
metal = { git = "https://github.com/gfx-rs/metal-rs.git", rev = "032800fe731082ff115a5346b5178d12cfd1bf02" }
```

Metal now compiles, but `gfx-backend-metal v0.2.4` fails with the same E0282 errors.

**Comment by 0x6273** [comment]
@trsoluti can you try with this version of gfx-backend-metal?
https://github.com/0x6273/gfx/tree/metal_v0.2

**Comment by trsoluti** [comment]
@0x6273 Doesn't appear to be possible to do this. The version of gfx-backend-metal you reference refers to metal 0.15.0, but the metal patch you sent me is 0.15.1. Both amethyst_rendy and gfx-backend-hal specify dependencies right down to the patch level. Cargo won't let you substitute in those cases.

**Comment by 0x6273** [comment]
Neither rendy nor gfx-backend-hal seems to require an exact version in their `Cargo.toml`s.
https://github.com/0x6273/gfx/blob/0eadbc000b2ea01e90bd73faa65b6398e01ef47a/src/backend/metal/Cargo.toml#L32
https://github.com/amethyst/rendy/blob/b2450fbbd72866f0c9f13dbdda02885a8a30d8f3/util/Cargo.toml#L31

I'm guessing cargo just doesn't like that the version number is different from the version on crates.io. Maybe you can try using just the bugfix commit without the version bump with `rev = "..."`?
https://github.com/0x6273/metal-rs/commit/61c3f662ece7f383030ead3fce116c82d9275b32
https://github.com/0x6273/gfx/commit/760f2d69dc46a1abf8beafdc096819910200a621

**Comment by ishitatsuyuki** [comment]
I think 0.14 is now available so we could close this?

**Comment by 0x6273** [comment]
> I think 0.14 is now available so we could close this?

I think we probably want to support at least the last few releases.

**Comment by ishitatsuyuki** [comment]
> I think we probably want to support at least the last few releases.

Well, I don't really think we have the time to do so. We're on semver 0.x so no long term support should be expected either.

**Comment by trsoluti** [comment]
I'd be quite happy to close this problem report. I agree with @ishitatsuyuki that we're in 0.x so no need to support previous releases.

**Comment by trsoluti** [comment]
Closing issue.

**Comment by luojia65** [comment]
What should macOS developer do to at least compile amethyst by now?

**Comment by trsoluti** [comment]
@luojia65, the out-of-box v0.14.0 version of Amethyst compiles fine on a Mac, as long as you choose --features metal. The graphics are smooth and don't fail, at least for my small space_shooter program. The only issue is the lack of UI graphics, for which I've submitted a bug report.

If you want to fix 0.12 or 0.13, I'm not sure what to do. The various suggestions given above don't work. Do you want me to re-open this problem report, specifically for fixing in 0.12 or 0.13?

**Comment by luojia65** [comment]
Oh yes updating to 0.14 solved my problem. I'd prefer modifying my game code to adapt to latest version. Thanks! @trsoluti
